feat(help): Add styling to help output

epage opened the pull request:
### What does this PR try to resolve?

Try to make `--help` output easier to parse by using terminal styling

Screenshots:

![Screenshot from 2023-09-06 09-57-11](https://github.com/rust-lang/cargo/assets/60961/61069af4-ef05-40ad-9240-fedea44d4c71)

![Screenshot from 2023-09-06 09-57-21](https://github.com/rust-lang/cargo/assets/60961/d2e69024-42aa-47c0-ad0f-24e43551b8db)

![Screenshot from 2023-09-06 09-57-36](https://github.com/rust-lang/cargo/assets/60961/e3d895e2-745f-48c6-9e84-d6fb67198d6d)

*(`nargo` is my shell script wrapping `cargo run --manifest-path cargo/Cargo.toml`)*

### How should we test and review this PR?

At this time, the only styling snapshotting library I know of is a pain to use, so testing this requires manually running the commands which I did.  Screenshots are included for easier evaluation of the general idea.

Snapshotting of the plain text output ensures we don't have accidental formatting regressions from this change since the formatting isn't as obvious from looking at the code.

### Additional information

Traditionally, cargo has disabled clap's styled output.  My assumed
reason is that cargo mixes custom help output with auto-generated and
you couldn't previously make it all styled.
Clap 4.2 allowed users to pass in strings styled using ANSI escape
codes, allowing us to pass in styled text that matches clap, unblocking this.  In clap
4.4.1, clap gained the ability for the user to override the style.

In this PR, I decided to use the new 4.4.1 feature to style clap's
output to match the rest of cargo's output.  Alternatively, we could use
a more subdue style that clap uses by default.

I used the `color-print` crate to allow something almost html-like for styling `&static str`.  Alternatively, we could directly embed the ANSI escape codes harder to get write, harder to inspect), or we could do the styling at runtime and enable the `string` feature in clap.

I decided to *not* style `Arg::help` messages because
- It might be distracting to have the descriptions lit up like a
  christmas tree
- It is a lot more work

The one exception I made was for `--list` since it is for a
psuedo-command (`...`) and I wanted to intentionally draw attention to
it.

#12593 made styling of `cargo -h` cleaner imo.
#12592 and #12594 were improvements I noticed while doing this.

epage commented:
Before applying this to the rest of cargo's commands, I figured I'd post this draft for feedback from the cargo team on (1) if we want this and (2) if we should make any tweaks

bors commented:
:umbrella: The latest upstream changes (presumably #12593) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

epage commented:
@rustbot ready

epage commented:
The one thing I'm not too sure of is if "literal" (bold blue) and "placeholder" (cyan) will work together in a variety of themes.  With my theme, it works well.  One workaround is to make them the same color (blue or cyan) with different intensity (bold or not) since themes would ideally treat those similarly.

bors commented:
:umbrella: The latest upstream changes (presumably #12607) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #12610) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #12609) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

weihanglo commented:
[zuilp discussion](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Styling.20help.20output.3F) of this PR

weihanglo reviewed (commented):
I am not entirely sure the choice of color. Do you have any reference for terminal styling convention? Should we write down Cargo's convention somewhere?

weihanglo commented on Cargo.lock:
```diff
@@ -541,6 +542,27 @@ version = "0.6.1"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "d51beaa537d73d2d1ff34ee70bc095f170420ab2ec5d687ecd3ec2b0d092514b"
+dependencies = [
+ "nom",
```
Just realized cargo depends on `nom` again (via `gix`).

epage replied:
I've removed `nom` from gix

weihanglo reviewed (approved):
This looks nice. One thing may need to address is that we now have `termcolor`, `color-print` and `clap` handling terminal styling in different ways in Cargo. Should we have a cleanup?

epage commented:
> I am not entirely sure the choice of color. Do you have any reference for terminal styling convention? Should we write down Cargo's convention somewhere?

Do you mean for us internally?  In most cases, it doesn't matter as its centrally defined in the clap code / `Shell` and the `color_print` code is mostly to copy the `clap` code

Do you mean publicly for others to copy?  That could be interesting.  One thought I had was to expose the style in `clap-cargo`

> This looks nice. One thing may need to address is that we now have `termcolor`, `color-print` and `clap` handling terminal styling in different ways in Cargo. Should we have a cleanup?

Don't forget `fwdansi` :)

I'm working on a bug fix to make it really solid but my plan was to migrate `cargo` from `termcolor` to `anstream` which is what `clap` uses under the hood.  Its more like `fwdansi` with less overhead so you then use it with some terminal styling crate like `color-print` for html-like formatting, `owo-colors` for a fancy API or `anstyle` for barebones styling.

If you mean centralizing the definitions of our colors, that is something that we can do.

- clap uses `anstyle`
- `anstyle-termcolor` exists for us to take a central `anstyle` definition and convert it to `termcolor` in `Shell` (until we use `anstream`)
- We could switch away from `color-print` to do runtime string formatting which would allow us to centralize the color definitions completely
  - `anstyle` isn't the most ergonomic, so we'd probably want to use `anstyle-owo-colors` or something
  - We'd need to do `cargo add clap -F string` to support runtime generated strings.

epage commented:
Decided to try out a variety of themes

<details>

"GNOME"

![Screenshot from 2023-09-05 11-57-48](https://github.com/rust-lang/cargo/assets/60961/cdcc1c8a-f712-4d81-a90a-e5e937fc48ec)

Tango

![Screenshot from 2023-09-05 11-59-03](https://github.com/rust-lang/cargo/assets/60961/9a86d9e8-dfb5-4555-8d6f-df74749b21ac)

"Linux console"

![Screenshot from 2023-09-05 11-59-30](https://github.com/rust-lang/cargo/assets/60961/bef6b23a-9e3c-40f2-8b88-61d93bfffa84)

"XTerm"

![Screenshot from 2023-09-05 11-59-53](https://github.com/rust-lang/cargo/assets/60961/5ac3dfca-eab2-4230-b586-f46a63964a3e)

"rxvt"

![Screenshot from 2023-09-05 12-00-18](https://github.com/rust-lang/cargo/assets/60961/b5c0fd05-7a68-48a0-998c-e932f941dac4)

Solarized

![Screenshot from 2023-09-05 12-00-37](https://github.com/rust-lang/cargo/assets/60961/72b8a65a-a33a-4111-8165-8b85460548bc)

I've also done a run through A-E of https://github.com/Gogh-Co/Gogh (thats a bit much to include screenshots for) and "XTerm" is likely the worst.

</details>

weihanglo commented:
Fantastic! You've thought way more thoroughly than I have 😂

I was initially thinking of writing down our colour palette for internal use, and then moving forward by centralizing the colour definitions. It would be great to have definitions for Cargo plugin developers to follow, even if we don't announce it in public.

I see most of the approaches you mentioned needing experiments. Perhaps we should have an issue tracking the entire progress.

---

Ooof xterm is pretty bad.  Is it one of the major themes? Might be a blocker to me if it is the case.

weihanglo commented:
(not a blocker)

The other issue is that this patch introduces one more dependency (and transitive proc-macro). This doesn't really affect Cargo development too much. However still some cargo plugins depend on cargo directly, so we still need to keep an eye on that.

epage commented:
> Ooof xterm is pretty bad. Is it one of the major themes? Might be a blocker to me if it is the case.

I switched it o Cyan + Bold.  In doing so I did a double check and realized that Yellow and Red were also supposed to be Bold, so I fixed that and updated the main screenshots

For other themes:

<details>

Gnome

![Screenshot from 2023-09-06 10-01-30](https://github.com/rust-lang/cargo/assets/60961/790a78b3-2371-4529-85ee-4e1cdbca2c78)

Tango

![Screenshot from 2023-09-06 10-01-37](https://github.com/rust-lang/cargo/assets/60961/c3ea09b3-1ff5-45e6-81bb-1096a7c14f6b)


Linux Console
![Screenshot from 2023-09-06 10-02-06](https://github.com/rust-lang/cargo/assets/60961/acb120a4-d772-4fb3-b689-1bfa176e4375)



Solarized


![Screenshot from 2023-09-06 10-02-20](https://github.com/rust-lang/cargo/assets/60961/a78a9b12-2084-46cf-bc93-239ca03fec96)

</details>

epage commented on Cargo.toml:
```diff
@@ -30,6 +30,7 @@ cargo-test-support = { path = "crates/cargo-test-support" }
 cargo-util = { version = "0.2.6", path = "crates/cargo-util" }
 cargo_metadata = "0.17.0"
 clap = "4.4.2"
+color-print = "0.3.4"
```
> (not a blocker)
>
> The other issue is that this patch introduces one more dependency (and transitive proc-macro). This doesn't really affect Cargo development too much. However still some cargo plugins depend on cargo directly, so we still need to keep an eye on that.

syn, quote, and proc-macro2 will already be pulled in.  I can see if I can work with the maintainer to switch to syn 2 (a gix dep also pulls in syn v1) and winnow (gix pulls in nom for now).  That leaves just the macro itself which shouldn't be all that much.  While is only used in the bin for now, I'd like to work towards styled text in more places once we are using anstream.  The main issue with color-print itself is theming.  This is also balanced out with not needing to pull in the `String` feature into clap which affects binary size, compile times, and runtime performance.

So for me, I'm not too worried about it and I can easy see the situation evolving.  To what?  I don't yet know.

epage commented:
> I see most of the approaches you mentioned needing experiments. Perhaps we should have an issue tracking the entire progress.

For at least part of it, I've created #12627

joshtriplett commented:
It feels *mildly* unfortunate that we have the concept of what "cyan" and "bold" semantically represent embedded in multiple places. This should *not* stop trying this experiment, but long-term I'm wondering how easily we might be able to introduce some compile-time semantic names for these styles, such as in a common wrapper around `cstr!`.

joshtriplett commented:
@epage Could you please provide sample screenshots for this color scheme on a black background, to verify that it doesn't create any hard-to-read text?

epage commented:
@joshtriplett see my earlier comment: https://github.com/rust-lang/cargo/pull/12578#issuecomment-1708552335

In addition to those screenshots, I've done a run through to spot check A-E of https://github.com/Gogh-Co/Gogh.

epage commented:
To add, all of the colors used are colors already used in cargo *except* the Cyan used in placeholders (e.g. `<command>`).  That was intentional both to both match cargo's existing colors and with the assumption that cargo's existing colors seem to be working out for us.  The Cyan for placeholders is just a variant of Cyan+Bold we already use.

joshtriplett commented:
@epage LGTM! Thank you.

@rfcbot reviewed

weihanglo commented:
Let's merge it and see how people feel about it 🚀 

@bors r+

bors commented:
:pushpin: Commit dbbc5dd6eba308f266fe67190cd04c27109dd736 has been approved by `weihanglo`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=weihanglo dbbc5dd6eba308f266fe67190cd04c27109dd736 -->
<!-- homu: {"type":"Approved","sha":"dbbc5dd6eba308f266fe67190cd04c27109dd736","approver":"weihanglo","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit dbbc5dd6eba308f266fe67190cd04c27109dd736 with merge c2d26f32d33ac0135183e141f5fb10955a344145...
<!-- homu: {"type":"BuildStarted","head_sha":"dbbc5dd6eba308f266fe67190cd04c27109dd736","merge_sha":"c2d26f32d33ac0135183e141f5fb10955a344145"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/6142733831/job/16666148326)
Approved by: weihanglo
Pushing c2d26f32d33ac0135183e141f5fb10955a344145 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"weihanglo","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/6142733831/job/16666148326"},"merge_sha":"c2d26f32d33ac0135183e141f5fb10955a344145"} -->

your-diary commented:
How can I disable this or change the used colors?

`--color never` didn't work: `cargo --color never build --help`

epage commented:
Could you go into why you want to disable it? It matches existing cargo colors so we thought it'd be innocuous. It'd be good to know what problems people are having with it.

As for disabling, the cli doesn't work because help gets processed before we can process `--color`. I can't remember if I hacked in `term.color` / `CARGO_TERM_COLOR` support but that'd be the next thing id recommend. Otherwise, as an implementation detail, we do support `NO_COLOR=1` (I think there is an underscore in there).

your-diary commented:
@epage 

> Could you go into why you want to disable it?

Because personally I don't like the colors used.
Additionally I believe there are not small number of people who want to opt-out the colorful output entirely, taking into account how many people in the world use Rust.
How colors look to one's eyes highly depends on the person.
As personal preferences of course significantly vary from person to person, I expect the command lets users customize its appearance as much as possible.
I believe this is ***one*** of the reasons why many CUI commands have `--color={auto|never|always}` option.
Some commands like `ls` even lets users customize which colorset is used (i.e. `LS_COLORS`).

> It matches existing cargo colors

I don't think so, partly. The first impression for the current output I had was "too bright".

Actually, as you say, the color cyan has been used in Cargo. But, as far as I know, the main color is green and cyan is just temporarily displayed for highlighting purposes:

<img width="492" alt="Screenshot 2023-11-18 at 22 30 37" src="https://github.com/rust-lang/cargo/assets/58146437/414e6058-ea29-4b7a-8778-28325b7ca73e">

However, in the output of `--help`, cyan is used as one of the main colors:

<img width="669" alt="Screenshot 2023-11-18 at 22 31 02" src="https://github.com/rust-lang/cargo/assets/58146437/429c9e10-f1ff-40c9-a859-6c984ced9c18">

This looks too bright for me.

> As for disabling, the cli doesn't work because help gets processed before we can process --color.

I see. I suspect so. I think this is not a good UI/UX; it seems at least the doc should be fixed. The current doc is this (`cargo --help` and `man cargo` respectively):

```
      --color <WHEN>            Coloring: auto, always, never
```

```
       --color when
           Control when colored output is used. Valid values:

           •  auto (default): Automatically detect if color support is available on
               the terminal.

           •  always: Always display colors.

           •  never: Never display colors.

           May also be specified with the term.color config value
           <https://doc.rust-lang.org/cargo/reference/config.html>.
```

>  I can't remember if I hacked in term.color / CARGO_TERM_COLOR support but that'd be the next thing id recommend.

Didn't work either. I'd be great if the doc would be fixed for the same reason described right above.

> Otherwise, as an implementation detail, we do support NO_COLOR=1

Worked like a charm. Thank you.
```bash
$ NO_COLOR=1 cargo build --help
```

weihanglo commented:
Thanks for calling them out. Some relevant contexts here about why it is hacky to support the option in CLI options and `CARGO_TERM_COLOR`:

* https://github.com/rust-lang/cargo/issues/2290
* https://github.com/rust-lang/cargo/issues/9012
* https://github.com/rust-lang/cargo/issues/11211

That said, this is a thing I would like to see it fixed.

epage commented:
> @epage 
> 
> > Could you go into why you want to disable it?
> 
> Because personally I don't like the colors used.

Note that we are using terminal-configured colors, rather than something like RGB, so you do have full control over how this gets rendered in your terminal.

your-diary commented:
@epage 
Yes. But the setting is global; it cannot (or maybe difficult if not impossible) be set on a per-command basis.

Strictly speaking, the problem is not the color itself. I feel some inconsistency in how the color is used:

> Actually, as you say, the color cyan has been used in Cargo. But, as far as I know, the main color is green and cyan is just temporarily displayed for highlighting purposes:
> 
> However, in the output of --help, cyan is used as one of the main colors:

Anyway, the essential problem is *all* of the Rust users with different personal preferences are now forced to be exposed to the colorful output.
`cargo build --color never` works. `cargo build --help --color never` does not work.

epage commented:
> Strictly speaking, the problem is not the color itself. I feel some inconsistency in how the color is used:

The reason I chose those colors for the help is that Green is used as more of a primary header while Cyan is a lower header / note. 

> cargo build --color never works. cargo build --help --color never does not work.

Usually a command-line flag is for transient behavior.  If this is for user preference, a CLI flag wouldn't be appropriate.

As there are related issues for this., I'd recommend we shift the focus to the relevant ones as anything discussed here will be lost track of.

your-diary commented:
@epage 

> Usually a command-line flag is for transient behavior. If this is for user preference, a CLI flag wouldn't be appropriate.

I don't think so. This is also a personal preference for design, I think.
I've learned 500 or more UNIX commands and many of them use CLI flags for user preferences.
And..., at least, `cargo` does support CLI flag for user preference:
```bash
function cargo() {
    if [[ $1 == 'build' ]]; then
        command cargo build --color never "$@"
    else
        command cargo "$@"
    fi
}
```

whether or not this is the way how `--color` is intended to be used.

> As there are related issues for this., I'd recommend we shift the focus to the relevant ones as anything discussed here will be lost track of.

I do agree.
I just subscribed to #9012.
As #9012 is more essential, I will suspend my discussion and won't create a new issue.
For the time being, I'll use `NO_COLOR=1`.

Thank you for the series of the replies.
